datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -1363,69 +1363,278 @@ pub fn array_ndims(args: &[ArrayRef]) -> Result<ArrayRef> {
+                if let (Some(arr), Some(elem)) = (arr, elem) {
+                    let list_arr = arr.as_list::<i32>();
+                    let res = list_arr.iter().dedup().flatten().any(|x| *x == *elem);
+                    boolean_builder.append_value(res);
+                }
+            }
+            Ok(Arc::new(boolean_builder.finish()))
+        }
+
+        // Int64, Int32, Int16, Int8
+        // UInt64, UInt32, UInt16, UInt8
+        DataType::Int64 => {
+            non_list_contains!(array, args[1], Int64Array)
+        }
+        DataType::Int32 => {
+            non_list_contains!(array, args[1], Int32Array)
+        }
+        DataType::Int16 => {
+            non_list_contains!(array, args[1], Int16Array)
+        }
+        DataType::Int8 => {
+            non_list_contains!(array, args[1], Int8Array)
+        }
+        DataType::UInt64 => {
+            non_list_contains!(array, args[1], UInt64Array)
+        }
+        DataType::UInt32 => {
+            non_list_contains!(array, args[1], UInt32Array)
+        }
+        DataType::UInt16 => {
+            non_list_contains!(array, args[1], UInt16Array)
+        }
+        DataType::UInt8 => {
+            non_list_contains!(array, args[1], UInt8Array)
+        }
+
+        DataType::Float64 => {
+            non_list_contains!(array, args[1], Float64Array)
+        }
+        DataType::Float32 => {
+            non_list_contains!(array, args[1], Float32Array)
+        }
+        DataType::Utf8 => {
+            non_list_contains!(array, args[1], StringArray)
+        }
+        DataType::LargeUtf8 => {
+            non_list_contains!(array, args[1], LargeStringArray)
+        }
+        DataType::Boolean => {
+            non_list_contains!(array, args[1], BooleanArray)
+        }
+        _ => {
+            todo!(
+                "array_has not implemented for type: {:?}",
+                args[1].data_type()
+            )
+        }
+    }
+}
 
+macro_rules! array_has_any_non_list_check {
+    ($ARRAY:expr, $SUB_ARRAY:expr, $ARRAY_TYPE:ident) => {{
+        let arr = downcast_arg!($ARRAY, $ARRAY_TYPE);
+        let sub_arr = downcast_arg!($SUB_ARRAY, $ARRAY_TYPE);
+
+        let mut res = false;
+        for elem in sub_arr.iter().dedup() {
+            res |= arr
+                .iter()
+                .dedup()
+                .flatten()
+                .any(|x| x == elem.expect("null type not supported"));
+        }
         res
     }};
 }
 
-/// Array_contains SQL function
-pub fn array_contains(args: &[ArrayRef]) -> Result<ArrayRef> {
-    fn concat_inner_lists(arg: ArrayRef) -> Result<ArrayRef> {
-        match arg.data_type() {
-            DataType::List(field) => match field.data_type() {
-                DataType::List(..) => {
-                    concat_inner_lists(array_concat(&[as_list_array(&arg)?
-                        .values()
-                        .clone()])?)
+/// Array_has_any SQL function
```

Comment:
Did not find a way that combines `has_any` and `has_all` without losing readability.